hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/key/OMKeysRenameRequest.java

```diff
@@ -0,0 +1,279 @@
+      bucket.audit(auditMap);
+      volumeName = bucket.realVolume();
+      bucketName = bucket.realBucket();
+
+      for (RenameKeysMap renameKey : renameKeysArgs.getRenameKeysMapList()) {
+
+        fromKeyName = renameKey.getFromKeyName();
+        toKeyName = renameKey.getToKeyName();
+        RenameKeysMap.Builder unRenameKey = RenameKeysMap.newBuilder();
+
+        if (toKeyName.length() == 0 || fromKeyName.length() == 0) {
+          renameStatus = false;
+          unRenamedKeys.add(
+              unRenameKey.setFromKeyName(fromKeyName).setToKeyName(toKeyName)
+                  .build());
+          LOG.error("Key name is empty fromKeyName {} toKeyName {}",
+              fromKeyName, toKeyName);
+          continue;
+        }
+
+        try {
+          // check Acls to see if user has access to perform delete operation
+          // on old key and create operation on new key
+          checkKeyAcls(ozoneManager, volumeName, bucketName, fromKeyName,
+              IAccessAuthorizer.ACLType.DELETE, OzoneObj.ResourceType.KEY);
+          checkKeyAcls(ozoneManager, volumeName, bucketName, toKeyName,
+              IAccessAuthorizer.ACLType.CREATE, OzoneObj.ResourceType.KEY);
+        } catch (Exception ex) {
+          renameStatus = false;
+          unRenamedKeys.add(
+              unRenameKey.setFromKeyName(fromKeyName).setToKeyName(toKeyName)
+                  .build());
+          LOG.error("Acl check failed for fromKeyName {} toKeyName {}",
+              fromKeyName, toKeyName, ex);
+          continue;
+        }
+
+        // Check if toKey exists
+        String fromKey = omMetadataManager.getOzoneKey(volumeName, bucketName,
+            fromKeyName);
+        String toKey =
+            omMetadataManager.getOzoneKey(volumeName, bucketName, toKeyName);
+        OmKeyInfo toKeyValue = omMetadataManager.getKeyTable().get(toKey);
+
+        if (toKeyValue != null) {
+
+          renameStatus = false;
+          unRenamedKeys.add(
+              unRenameKey.setFromKeyName(fromKeyName).setToKeyName(toKeyName)
+                  .build());
+          LOG.error("Received a request name of new key {} already exists",
+              toKeyName);
+        }
+
+        // fromKeyName should exist
+        fromKeyValue = omMetadataManager.getKeyTable().get(fromKey);
+        if (fromKeyValue == null) {
+          renameStatus = false;
+          unRenamedKeys.add(
+              unRenameKey.setFromKeyName(fromKeyName).setToKeyName(toKeyName)
+                  .build());
+          LOG.error("Received a request to rename a Key does not exist {}",
+              fromKey);
+          continue;
+        }
+
+        fromKeyValue.setUpdateID(trxnLogIndex, ozoneManager.isRatisEnabled());
+
+        fromKeyValue.setKeyName(toKeyName);
+
+        //Set modification time
+        fromKeyValue.setModificationTime(Time.now());
+
+        acquiredLock =
+            omMetadataManager.getLock().acquireWriteLock(BUCKET_LOCK,
```

Comment:
I think here we should acquire lock and release lock at end of the operation. 

Lets say this thread checked the fromKey and toKey and other delete thread acquire the lock and deleted the fromKey and this thread waiting for lock, once delete completes we update the cache and rename it to toKey.

So, now we renamed a deleted Key. 

And also there can be other scenarios like commitKey committed toKey and this thread assumes there is no such Key.

To avoid these kinds of scenarios check of Key from Table and add to Response should be done holding lock.

Reply:
Thank @bharatviswa504 . Agreed and  updated  this PR  to fix the problem.